### DISCUSSION TITLE
Typing and text display in right-to-left languages such as Persian (Farsi), Arabic, etc.

### DISCUSSION THREAD

**Discussion by aaa1386** [freeplane/freeplane#1425] [Ideas]
Some users use Farsi or Arabic or other languages that operate on the basis of right to left in the mind map.
  Some errors occur in different areas of Freeplane, such as note panel or node editing.
For example, the home and end keys do not work well when editing. Or numerical list or bullet list option
In the video below, the bugs are shown
In the pictures, the correct mode is shown in right-to-left languages in Microsoft Word software.

⬇️⬇️⬇️A mind map with sample characters of a language that is based on right to left

[example freeplane file.zip](https://github.com/freeplane/freeplane/files/12658068/example.freeplane.file.zip)

❗️❗️❗️❗️❗️To better understand the errors, I recommend installing the Arabic or Farsi (Persian) language on Windows and testing.

In this video, when I am typing, I only put a space after each word, See irregular movements of words⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️

https://github.com/freeplane/freeplane/assets/38586259/63ec18c4-0624-4777-81fc-9a0fb97e10bd

Other errors in typing and display⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️

https://github.com/freeplane/freeplane/assets/38586259/8be10e0c-1db8-4739-b117-a8d38dc3b3dc

The correct mode is shown in right-to-left languages in Microsoft Word software.☑️☑️☑️☑️☑️☑️☑️☑️☑️  ⬇️⬇️⬇️⬇️⬇️

![image](https://github.com/freeplane/freeplane/assets/38586259/9900f315-ead4-40aa-826a-a09d78ad3650)
![image](https://github.com/freeplane/freeplane/assets/38586259/3efc3dcf-b55f-4f93-967c-25f505126216)

**Comment by theelectricsoul** [comment]
@ajaberi1360 
I tested it. The cursor position is jumping in freeplane but also in word.

Have you tried changing the language in Freeplane?

**Comment by aaa1386** [comment]
Yes, I changed the language of the program from the settings section to Arabic, which is based on right to left, but the errors are still there.
In fact, these errors are related to the difference between right-to-left or left-to-right languages. The Freeplane software focuses mainly on left-to-right languages such as English. I hope the developers of this software will correct these errors

**Comment by dpolivaev** [comment]
If Freeplane does not fully support right-to-left languages, it is neither a bug nor an error but a not implemented feature.

**Comment by aaa1386** [comment]
Thank you .
I created this thread to help make the good Freeplane software more complete.
If it is not possible to fix the reported defects, it is not a problem, I use the software in the same situation.
But if you can correct the mentioned items, the software will be more complete and users are more comfortable to use
Thank you again for your efforts

**Comment by theelectricsoul** [comment]
@ajaberi1360 if dpolivaev personally adds a "make sense" tag to a request, you are good :) And he did :)   From my experience we just have to wait. 

Making it work for arabic languages seems like a very sensible goal to reach, so ... I am confident it will be implemented. Depends on the effort.

**Comment by aaa1386** [comment]
@theelectricsoul 
I checked that 25 countries in the world speak Arabic. The statistics are significant.
  There are other languages such as Farsi that are based on right to left.
Maybe there are other languages, I didn't check more than that.
  
I don't know whether Freeplane developers pay attention to such languages or not. Whatever they decide, I respect their decision.

**Comment by euu2021** [comment]
In sourceforge, it's possible to see statistcs and maps by number of downloads: 
https://sourceforge.net/projects/freeplane/files/stats/map

Saudi Arabia is in the all time Top10:
![image](https://github.com/freeplane/freeplane/assets/77707706/6fcb2fe7-f083-43ba-9d59-1c057650de52)

**Comment by dpolivaev** [comment]
I care about everyone, so I checked what can be done for better support of right-to-left languages.
I found that this option can be configured for each node, detail, or note, but can not be changed inside one node, detail, or note.

Therefore I am going to add a node format checkbox "right-to-left", with a default value coming from the language set as system locale and also configurable as a Freeplane user setting.

At the moment I convert this issue to a discussion to get wider feedback from the community about this approach.

**Reply by aaa1386** [reply]
dpolivaev@ Thank you for following up
Can you explain a little more clearly?
Do you mean that by changing the language, for example, from English to Arabic, the necessary changes do not happen automatically?
And should we check the right to left check box for changes and correct writing of the text?

**Reply by dpolivaev** [reply]
Yes, you got it right.

**Comment by aaa1386** [comment]
The idea you gave is great if it is possible

**Comment by dpolivaev** [comment]
Preview 1.11.8_01

**Comment by aaa1386** [comment]
Thank you for your effort.
But nothing special has happened and it is not working well.
I did a comparison between Xmind software and Freeplane, maybe it will help.
The XMind software does not work correctly in the node and sub-branches, but it works accurately in the note section



https://github.com/freeplane/freeplane/assets/38586259/6fe95000-4530-4ed5-8837-954aa3371a40

**Reply by dpolivaev** [reply]
1. Comparison with XMind is not relevant.
2. Something has happened: now you can configure text direction for each node / node style, and when you do so the settings are also applied to notes and details attached to the node.

For configuring text direction you can use UI elements indicated on the screenshot.

<img width="670" alt="testDirection" src="https://github.com/freeplane/freeplane/assets/1558052/9bbe60a8-667d-4d65-9f9c-a7f039c7b6c3">

**Comment by aaa1386** [comment]
If you see the video and free plan files below, you will find the exact and final solution, maybe by combining the idea you implemented with my idea, the work will be completed.
  See, when I manually edit the numbers, the numbers are written in Arabic or Farsi, and their structure shows correctly without using the left-to-right or right-to-left icon.
Therefore, the main problem is the lack of numbering in Arabic or Persian language. If you pay attention, Microsoft Word has prepared two types of numbering, one in English and one in Arabic or Persian. Shown inside the video
Therefore, you should be in the edit node core in dialog
  Design two types of numbering, both English and Arabic or Persian.


https://github.com/freeplane/freeplane/assets/38586259/634ab020-e4e9-46e9-a5e3-9930a4e14024

https://github.com/freeplane/freeplane/assets/38586259/0dd2a5e9-d2f5-41a6-b2b2-1a61c94abf69

**Comment by aaa1386** [comment]
![image](https://github.com/freeplane/freeplane/assets/38586259/2f80bf1a-e28d-481d-af2a-f80692a92f7f)
⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️⬇️

**Comment by aaa1386** [comment]
⬇️⬇️⬇️⬇️⬇️⬇️
[New Mindmap.zip](https://github.com/freeplane/freeplane/files/12810160/New.Mindmap.zip)

**Comment by aaa1386** [comment]
⬇️⬇️⬇️⬇️⬇️⬇️
[IRNumber.zip](https://github.com/freeplane/freeplane/files/12810221/IRNumber.zip)

font

**Comment by aaa1386** [comment]
![image](https://github.com/freeplane/freeplane/assets/38586259/5714345f-d32f-4894-81e2-f4270c48c22c)
The rule of making Arabic and English numbers is the same

**Comment by dpolivaev** [comment]
Unfortunately, your suggestions can not be implemented because the used library does not support them.
If other tools fit your needs better, you can use them.
If the solution I implemented is not helpful, let me know, so that I can consider reverting the change.

**Comment by aaa1386** [comment]
Thanks
Yes, I understand that not all ideas may be applicable
My opinion is that the implemented measures should not be removed, these changes are a step forward.
   This is my opinion, others may have a different opinion

**Comment by aaa1386** [comment]
Maybe someday the idea will be more complete

**Comment by aaa1386** [comment]
In the new update, the requested item works much better.
thanks.

Typing in "edit note core in dilog" works very well, but typing in "edit node core in-line" is not as good as before. However, in the end, there is nothing wrong with the typed text.Anyway, the changes that have happened are remarkable and satisfying.
See the video below.
The indicator does not work well when home and end in non-English language.
When typing, the text skips, but after pressing enter, the text is displayed fine

https://github.com/freeplane/freeplane/assets/38586259/453e637a-c547-49e3-8e31-681adf60c30d

**Comment by aaa1386** [comment]
If possible, put text direction icon in the note panel.
In the note panel, it would be great if it works like "write node in the dialog"

**Reply by dpolivaev** [reply]
Describe a use case where you need independent control of the text direction in the node core, the details, the notes, and the attributes. Separate configuration would add complexity, particularly considering the styles, and I would like to avoid it.

**Reply by aaa1386** [reply]
The software is acceptable and does not cause any major problems.
You are right

**Reply by dpolivaev** [reply]
Do you actually understand that calling software "acceptable" sounds like disapproval?

**Reply by aaa1386** [reply]
I am sorry. I don't know English well and I don't know words well. I mean, the software is very good and the right-to-left language support is acceptable. I am satisfied with Freeplane software and love it.

**Comment by EdoFro** [comment]
> Do you actually understand that calling software "acceptable" sounds like disapproval?

🤣 Most of us learn these differences in lenguages the hard way!!!

It's like "I'm happy for you", that in chilean spanish is the usual acid way to say "I don't care about it (or about you) / I'm not interested in your opinion."
Every time I read this expression I have a mental conflict trying to understand if it was really meant or it was sarcasm. 🤣
